### PR TITLE
Remove most usages of *wprintf* family of functions

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Removing usage of the `wprintf` family of APIs comes down to `wchar_t` and what it means on what platform.

Since we use raw `wprintf` style functions, we can never use standard platform headers because most are defined with `wchar_t`, which is 4-bytes on most non-Windows platforms but 2-bytes on Windows. However, our usage of all these APIs are all based on the Windows definition and that is spread throughout the code base. This means we naively pass `WCHAR` through and it works. However, if we use the system platform headers they are defined with `wchar_t` and that assumes that platform's definition of `wchar_t` which may not be 2-bytes.

This is all in the service of cleaning up the PAL and eventually removing our private implementation of `wprintf`.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
@jkotas @janvorli @dotnet/dotnet-diag-contrib Please take a look at these changes. There are a few additional changes that are needed but with this PR we are down to a narrow set of cases to handle.

File: src/coreclr/vm/dispatchinfo.cpp [hunk 1]
```diff
@@ -1755,8 +1755,8 @@ void DispatchInfo::InvokeMemberWorker(DispatchMemberInfo*   pDispMemberInfo,
         // Allocate the string that will contain the name of the member.
         if (!pDispMemberInfo)
         {
-            WCHAR strTmp[64];
-            _snwprintf_s(strTmp, ARRAY_SIZE(strTmp), _TRUNCATE, DISPID_NAME_FORMAT_STRING, id);
+            UTF8 strTmp[ARRAY_SIZE(DISPID_NAME_FORMAT_STRING "4294967295")];
```
Comment (jkotas) [inline src/coreclr/vm/dispatchinfo.cpp:1758]: Nit: Windows specific. Can stay as is.

File: src/coreclr/vm/dwbucketmanager.hpp [hunk 2]
```diff
@@ -484,11 +484,10 @@ void BaseBucketParamsManager::GetAppVersion(_Out_writes_(maxLength) WCHAR* targe
 
     if ((GetCurrentModuleFileName(appPath) == S_OK) && SUCCEEDED(DwGetFileVersionInfo(appPath, major, minor, build, revision)))
     {
-        _snwprintf_s(targetParam,
-            maxLength,
-            _TRUNCATE,
-            W("%d.%d.%d.%d"),
-            major, minor, build, revision);
+        WCHAR* cur = FormatInteger(targetParam, maxLength, "%d.", major);
```
Comment (jkotas) [inline src/coreclr/vm/dwbucketmanager.hpp:487]: Nit: Windows-specific. Can stay as is.

**Review by jkotas — COMMENTED** [review]
Looks good to me overall. Thank you!

File: src/coreclr/debug/ee/debugger.cpp [hunk 3]
```diff
@@ -1735,15 +1735,15 @@ CLR_ENGINE_METRICS g_CLREngineMetrics = {
...
     &g_hContinueStartupEvent};
 
-#define StartupNotifyEventNamePrefix W("TelestoStartupEvent_")
-const int cchEventNameBufferSize = sizeof(StartupNotifyEventNamePrefix)/sizeof(WCHAR) + 8; // + hex DWORD (8).  NULL terminator is included in sizeof(StartupNotifyEventNamePrefix)
 HANDLE OpenStartupNotificationEvent()
 {
-    DWORD debuggeePID = GetCurrentProcessId();
-    WCHAR szEventName[cchEventNameBufferSize];
-    swprintf_s(szEventName, cchEventNameBufferSize, StartupNotifyEventNamePrefix W("%08x"), debuggeePID);
+    WCHAR prefix[] = W("TelestoStartupEvent_");
```
[the remark this answers is not in the document]
Comment (hoyosjs) [reply src/coreclr/debug/ee/debugger.cpp:1740]: It gets used for `NotifyDebuggerOfStartup` that the shim listens for at startup on Windows in https://github.com/dotnet/diagnostics/blob/13144c2fc43b714e924f1690f59427ad2782332c/src/dbgshim/dbgshim.cpp#L952-L986. Renaming it causes more grief than benefit.

[the remark this answers is not in the document]
Comment (noahfalk) [reply src/coreclr/debug/ee/debugger.cpp:1740]: Adding a comment might be a cheap alternative to reduce confusion.

File: src/coreclr/debug/ee/debugger.cpp [hunk 4]
```diff
@@ -6880,7 +6880,11 @@ HRESULT Debugger::EDAHelper(PROCESS_INFORMATION *pProcessInfo)
 
             LOG((LF_CORDB, LL_INFO10000, "D::EDA: launching with command [%S]\n", wszDbgCommand));
 
-            ClrGetCurrentDirectory(strCurrentDir);
+            // Get size needed
+            DWORD len = WszGetCurrentDirectory(strCurrentDir);
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/debug/ee/debugger.cpp:6884]: Any concerns @dotnet/dotnet-diag ?

File: src/coreclr/vm/commtmemberinfomap.cpp [hunk 5]
```diff
@@ -17,8 +17,15 @@
 
 #define BASE_OLEAUT_DISPID 0x60020000
 
-static LPCWSTR szDefaultValue           = W("Value");
-static LPCWSTR szGetEnumerator          = W("GetEnumerator");
+// Constants.
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/commtmemberinfomap.cpp:20]: I'd prefer to remove all instances when possible. This change appears simple to audit and I would prefer to remove all uses so there is no confusion about what is/isn't a concern.

File: src/coreclr/debug/daccess/daccess.cpp [hunk 6]
```diff
@@ -5730,6 +5730,87 @@ ClrDataAccess::GetJitHelperName(
...
+    _Out_writes_bytes_opt_(bufLen) WCHAR* symbolBuf)
+{
+    // Parts needed to construct a name:
+    // With stub manager name: "CLRStub[%s]@%p"
+    //   No stub manager name: "CLRStub@%p"
+    const WCHAR formatName_Prefix[] = W("CLRStub");
+    const WCHAR formatName_OpenBracket[] = W("[");
+    const WCHAR formatName_CloseBracket[] = W("]");
+    const WCHAR formatName_PrefixEnd[] = W("@");
+    const size_t formatName_AddrMaxLen = ARRAY_SIZE("ffffffffffffffff") - 1; // 64-bit
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/debug/daccess/daccess.cpp:5748]: It is hard to tell that this string is correct by just looking at it. If this string had 15 characters instead of 16, nobody will notice that during codereview.

File: src/coreclr/vm/commtmemberinfomap.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/vm/commtmemberinfomap.cpp:20]: Once you delete `wprintf*` from the Unix PAL, the compiler will audit this for you. There is no risk of missing one instance somewhere.

In general, we do not need to cleanup the Windows specific constructs that are problematic for portability from Windows specific code. It is just unnecessary work and risk.

[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/commtmemberinfomap.cpp:20]: Okay.

**Review by mikem8361 — APPROVED** [review]
The diagnostics look good as far as I can tell.

File: src/coreclr/utilcode/debug.cpp [hunk 7]
```diff
@@ -378,8 +389,8 @@ bool _DbgBreakCheck(
     // Emit assert in debug output and console for easy access.
     if (formattedMessages)
     {
-        WszOutputDebugString(debugOutput);
-        fwprintf(stderr, W("%s"), (const WCHAR*)debugOutput);
+        OutputDebugStringUtf8(debugOutput.GetUTF8NoConvert());
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/utilcode/debug.cpp:392]: I didn't realize that `SString::Printf` assumed ANSI. This is a bit messy then. Yes, I think we need `SString::PrintfUtf8`.

[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/coreclr/utilcode/debug.cpp:392]: This entire thing can be simplified. No need for a new UTF8 API.

File: src/coreclr/debug/ee/debugger.cpp [hunk 8]
```diff
@@ -1735,15 +1735,17 @@ CLR_ENGINE_METRICS g_CLREngineMetrics = {
...
 
-#define StartupNotifyEventNamePrefix W("TelestoStartupEvent_")
-const int cchEventNameBufferSize = sizeof(StartupNotifyEventNamePrefix)/sizeof(WCHAR) + 8; // + hex DWORD (8).  NULL terminator is included in sizeof(StartupNotifyEventNamePrefix)
 HANDLE OpenStartupNotificationEvent()
 {
-    DWORD debuggeePID = GetCurrentProcessId();
-    WCHAR szEventName[cchEventNameBufferSize];
-    swprintf_s(szEventName, cchEventNameBufferSize, StartupNotifyEventNamePrefix W("%08x"), debuggeePID);
+    // "Telesto" is the legacy name for the start up event. Changing this name
+    // would impact multiple repos and not worth the effort at present.
```
Comment (jkotas) [inline src/coreclr/debug/ee/debugger.cpp:1741]:
```suggestion
    // would impact multiple repos and not worth the effort at present.
    // This event is used by https://github.com/dotnet/diagnostics/blob/main/src/dbgshim/dbgshim.cpp
```

File: src/coreclr/vm/typestring.cpp [hunk 9]
```diff
@@ -387,10 +366,11 @@ HRESULT TypeNameBuilder::AddArray(DWORD rank)
         else             // allocation OK, do it the fast way
```
Comment (jkotas) [inline src/coreclr/vm/typestring.cpp:366]: This whole thing about trying to be robust in the presence of allocation failures is way overengineered. You can change this to just stackalloc the buffer while you are on it.

File: src/coreclr/vm/dwreport.cpp [hunk 10]
```diff
@@ -249,9 +249,14 @@ int DwGetAppDescription(                // Number of characters written.
     }
 
     // Build the query key for the language-specific file description resource.
-    WCHAR buf[64];                 //----+----1----+----2----+----3----+----4----+
-    _snwprintf_s(buf, ARRAY_SIZE(buf), _TRUNCATE, W("\\StringFileInfo\\%04x%04x\\FileDescription"),
-               translation->language, translation->codePage);
+    WCHAR buf[64];
```
Comment (jkotas) [inline src/coreclr/vm/dwreport.cpp:252]: Unnecessary change. This whole file is Windows specific.

**Review by jkotas — APPROVED** [review]
Thank you!
